faiss/IndexIVF.cpp

```diff
@@ -625,6 +629,53 @@ void IndexIVF::search_preassigned(
             for (int64_t i = 0; i < n; i++) {
                 reorder_result(distances + i * k, labels + i * k);
             }
+        } else if (pmode == 4) {
+            /// for this parallelization method, we parallelize over every
+            /// existing list, group together all the queries affiliated, and
+            /// scan all the queries of this list in one go.
+#pragma omp single
+            {
+                keys_by_list.resize(nlist);
+                for (idx_t i = 0; i < nlist; i++) {
+                    keys_by_list[i].reserve(nprobe * n / nlist);
+                }
+                for (idx_t il = 0; il < n; il++) {
+                    for (idx_t ikey = 0; ikey < nprobe; ikey++) {
+                        keys_by_list[keys[il * nprobe + ikey]].emplace_back(il);
+                    }
+                }
+                if (metric_type == METRIC_INNER_PRODUCT) {
+                    heap_heapify<HeapForIP>(k * n, distances, labels);
+                } else {
+                    heap_heapify<HeapForL2>(k * n, distances, labels);
+                }
+            }
+#pragma omp for schedule(dynamic)
+            for (idx_t il = 0; il < invlists->nlist; il++) {
+                std::vector<idx_t>& my_list = keys_by_list[il];
+                idx_t num_of_x = my_list.size();
+                if (num_of_x > 0) {
+                    std::vector<idx_t> local_idx(k * num_of_x);
+                    std::vector<float> local_dis(k * num_of_x);
```

Comment:
should be initialized with the neutral value (+ or - infinity)

Reply:
handled by the last commit.